teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -0,0 +1,198 @@
+---
+status: proposed
+title: Graceful Pipeline Run Termination
+creation-date: '2021-03-18'
+last-updated: '2021-03-18'
+authors:
+- '@@rafalbigaj'
+---
+
+# TEP-0058: Graceful Pipeline Run Termination
+---
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Marking a `PipelineRun` as [cancelled](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#cancelling-a-pipelinerun)
+stops all running tasks and deletes associated Pods. That prevents final tasks, specified in a `Pipeline` under `finally` section,
+from being run.
+There should be a way to terminate a `PipelineRun` gracefully and wait for cleanup actions triggered by final tasks.
+
+## Motivation
+
+It is common that tasks when run trigger execution of external activities or request resources in foreign systems.
+Final tasks are perfectly suited for any cleanup operations that have to be performed when execution is completed.
+That is required in case of success, failure, as well as pipeline run cancellation.
+ 
+Currently, there is no way to terminate a `PipelineRun` gracefully.
+The existing cancellation capability is problematic in the real use cases.
+
+This is especially important for users of [Kubeflow Pipelines with Tekton backend](https://github.com/kubeflow/kfp-tekton/).
+Kubeflow Pipelines supports [exit handler](https://www.kubeflow.org/docs/components/pipelines/overview/pipelines-overview/)
+which guarantees that selected operations are triggered whenever pipeline run is completed.
+Those actions are executed in case of a pipeline run cancellation as well. 
+Given the fact that Kubeflow Pipelines' `ExitHandler` is implemented in Tekton using final tasks,
+there is a significant inconsistency in the describe behaviour.
+
+Lastly, final tasks should be triggered on a pipeline run timeout, which is a standard error scenario.
+Running final tasks infinitely should be prevented with the additional configuration of a finalization timeout.
+There is a separate proposal: [TEP-0047](https://github.com/tektoncd/community/pull/326) that covers this part.
```

Comment:
Even if related, I think the behaviour of finally with respect to "cancel" and "timeout" are two separate issues, so having to distinct TEPs seems ok to me.

Reply:
@vincent-pli I agree on this additional requirement. Let's introduce it in baby steps.